### ISSUE TITLE
Discussion: Force extension methods for value types mark first argument as IN or REF 

### ISSUE THREAD

**Issue by maxkoshevoi** [dotnet/roslyn-analyzers#3535]
Take a look at following code:
```cs
struct Str
{
    public int Num;
}

static class Ex
{
    public static void Increment(this Str str)
    {
        str.Num++;
    }
}

var str = new Str();
str.Increment();
Console.WriteLine(str.Num); // Actual: 0, Expected: 1
```

`str` is copied by value, so any state changes made in `Increment` will be lost. 
Since C# 7.2 we have ability to [add the `ref` or `in` modifier to the first argument of an extension method](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods#extending-predefined-types).

- With `in` we explicitly saying that method won't change state of the object (and VS makes sure of that).
- With `ref` we explicitly saying that method will change state of the object and we need to save the changes.
- Without modifiers no changes will be saved and no warnings or errors will appear.
- I can see no downsides of adding `in` or `ref` in this scenario. Only benefits in form of additional compile-time checks 

Proposal: let's force all extension methods for value types to add `in` or `ref` to the first argument.

Possible duplicate of #2475 (proposal is a little different. if it's duplicate, I'll move discussion there)

**Comment by maxkoshevoi** [comment]
Here is an analyzer that could be used:

```cs
private void AnalyzeParameter(SymbolAnalysisContext context)
{
    var parameter = (IParameterSymbol)context.Symbol;

    if (parameter.ContainingSymbol is IMethodSymbol method
        && method.IsExtensionMethod
        && parameter.Ordinal == 0
        && parameter.Type.IsValueType 
        && parameter.RefKind == RefKind.None)
    {
        var diagnostic = Diagnostic.Create(Rule, context.Symbol.Locations[0], method.Name, parameter.Name);
        context.ReportDiagnostic(diagnostic);
    }
}
```

**Comment by maxkoshevoi** [comment]
@Evangelink WDYT?

**Comment by Evangelink** [comment]
That seems interesting but why not go one step further and apply this everywhere? We could use the `api_surface` user-option and the `excluded_symbol_names` if we want to allow some restriction.

**Comment by maxkoshevoi** [comment]
> apply this everywhere

What do you mean by that? Use this rule for all struct methods (even for non extension ones) or use this for all extension methods.

I don't think that using this rule for class extension method is a good idea. Take a look at "But there’s more…" section of [this post](https://blog.dunnhq.com/index.php/2017/12/15/readonly-parameters-in-c-a-step-closer-to-immutability/)

**Comment by Evangelink** [comment]
> > apply this everywhere
> 
> What do you mean by that? Use this rule for all struct methods (even for non extension ones) or use this for all extension methods.
> 
Not sure what I meant xD

I think the rule is a good idea, if the code targets C# 7.2+, the analyzer should ensure that when a `struct` is passed as first parameter of an extension method it should always have either `in` or `ref` to have a clear intent and avoid potential bugs.

@mavasani If you give us the green light I'd be happy to work on it (or let @maxkoshevoi if he wants to).

**Comment by mavasani** [comment]
I think @sharwell is more suitable to review this request, he has written similar analyzers for requiring `in` arguments.

**Comment by sharwell** [comment]
Passing by reference (even using `in`) disables certain optimizations in the JIT. I've observed cases where the cost of copying a value type is lower than the runtime cost of disabling these portions of the JIT, so it's not always valuable to write extension methods like this.

**Comment by maxkoshevoi** [comment]
@sharwell does passing value type with `in` still passes it by reference, and not by value? It seems like it should be passed by value both when nothing or `in` is used.

**Comment by sharwell** [comment]
`in` is the same as `ref`, except you can only read from it.

**Comment by maxkoshevoi** [comment]
@sharwell thank you for clarifying, I missed that somehow. How about this: analyzer that checks whether extension method for value type modifies that value type, and if it is, produces a warning to add `ref` to the first argument.

**Comment by maxkoshevoi** [comment]
@Evangelink, @mavasani @sharwell Hi again =)
So, what do you think about analyzer in the comment above? It still have the benefits of warning user about data loss and doesn't have drawbacks of disabling JIT optimizations without valid reason (user can refactor method so it doesn't change first argument instead of adding `ref`).

**Comment by sharwell** [comment]
> How about this: analyzer that checks whether extension method for value type modifies that value type  ...

I tried doing this, but it was much more difficult than expected. There were two primary problems:

1. Many types that are effectively read-only are marked as though they were mutable, and in many cases the analyzer can't tell the difference.
2. Many _members_ that could be marked `readonly` are not, and in many cases the analyzer can't tell the difference.

This led to an overwhelming number of false positives in analysis.
